File: sol/solution/0001-0100/0020/Solution.rb
# @param {String} s
# @return {Boolean}
def is_valid(s)
    stack = []
    s.each_char do |c|
        if c == '(' || c == '{' || c == '['
            stack.push(c)
        else
            return false if stack.empty?
            top = stack.pop
            return false if (c == ')' && top != '(') || 
                           (c == '}' && top != '{') || 
                           (c == ']' && top != '[')
        end
    end
    stack.empty?
end

File: sol/solution/0001-0100/0020/Solution.php
class Solution {
    /**
     * @param String $s
     * @return Boolean
     */
    function isValid($s) {
        $stack = [];
        foreach (str_split($s) as $c) {
            if ($c == '(' || $c == '{' || $c == '[') {
                array_push($stack, $c);
            } else {
                if (empty($stack)) return false;
                $top = array_pop($stack);
                if (($c == ')' && $top != '(') || 
                    ($c == '}' && $top != '{') || 
                    ($c == ']' && $top != '[')) {
                    return false;
                }
            }
        }
        return empty($stack);
    }
}